code/game/objects/items/weapons/rpd.dm

```diff
@@ -43,96 +23,126 @@
 	origin_tech = "engineering=4;materials=2"
 	var/datum/effect_system/spark_spread/spark_system
 	var/lastused
-	var/iconrotation = 0 //used to orient icons and pipes
-	var/mode = 1 //Disposals, atmospherics, etc.
-	var/pipetype = 1//For nanoUI menus
-	var/whatpipe = 0 //What kind of pipe is it? See code/game/machinery/pipe/construction.dm for a list of defines
-	var/whatdpipe = 0 //What kind of disposals pipe is it? See code/game/machinery/pipe/dispenser.dm for a list of defines
-	var/spawndelay = 4 //How long should we have to wait between dispensing pipes?
-	var/walldelay = 40 //How long should drilling into a wall take?
+	var/iconrotation = 0 //Used to orient icons and pipes
+	var/mode = RPD_ATMOS_MODE //Disposals, atmospherics, etc.
+	var/pipe_category = RPD_ATMOS_PIPING//For nanoUI menus, this is a subtype of pipes e.g. scrubbers pipes, devices
+	var/whatpipe = PIPE_SIMPLE_STRAIGHT //What kind of atmos pipe is it?
+	var/whatdpipe = PIPE_DISPOSALS_STRAIGHT //What kind of disposals pipe is it?
+	var/spawndelay = RPD_COOLDOWN_TIME
+	var/walldelay = RPD_WALLBUILD_TIME
 
 /obj/item/rpd/New()
 	..()
 	spark_system = new /datum/effect_system/spark_spread()
 	spark_system.set_up(1, 0, src)
 	spark_system.attach(src)
 
+/obj/item/rcd/Destroy()
+	QDEL_NULL(spark_system)
+	return ..()
+
 //Procs
 
-/obj/item/rpd/proc/Activaterpd(delay)
+/obj/item/rpd/proc/activate_rpd(delay) //Maybe makes sparks and activates cooldown if there is a delay
 	playsound(loc, "sound/machines/click.ogg", 50, 1)
 	if(prob(15))
 		spark_system.start()
 	if(delay)
 		lastused = world.time
 
-/obj/item/rpd/proc/Manipulatepipes(subject, atmosverb, disposalsverb)
-	if(istype(subject, /obj/item/pipe))
-		call(subject, atmosverb)()
-	else if(istype(subject, /obj/structure/disposalconstruct/))
-		call(subject, disposalsverb)()
+/obj/item/rpd/proc/can_dispense_pipe(var/pipe_id, var/pipe_type) //Returns TRUE if this is a legit pipe we can dispense, otherwise returns FALSE
+	for(var/list/L in GLOB.construction_pipe_list)
+		if(pipe_type != L["pipe_type"]) //Sometimes pipes in different categories have the same pipe_id, so we need to skip anything not in the category we want
+			continue
+		if(pipe_id == L["pipe_id"]) //Found the pipe, we can dispense it
+			return TRUE
+
+/obj/item/rpd/proc/create_atmos_pipe(mob/user, turf/T) //Make an atmos pipe, meter, or gas sensor
+	if(!can_dispense_pipe(whatpipe, RPD_ATMOS_MODE))
+		to_chat(user, "<span class='warning'>[src] beeps, \"ERROR\" </span>") //Damn dirty apes -- I mean hackers
```

Comment:
@tigercat2000 Do you want this to message admins in the event that this happens?

Reply:
`log_runtime(EXCEPTION("Failed to spawn pipe [whatpipe] - possible tampering detected"))` can work
Not `whatpipe` directly, of course, since that's just an ID that doesn't provide much info on its own